src/corelibs/U2Core/src/util/MsaRowUtils.cpp

```diff
@@ -432,4 +432,80 @@ void MsaRowUtils::removeTrailingGapsFromModel(qint64 length, QVector<U2MsaGap> &
     }
 }
 
+QByteArray MsaRowUtils::getGappedSubsequence(const U2Region &region, const QByteArray &sequence, const QVector<U2MsaGap> &gaps) {
+    CHECK(region.length > 0, {});
+    CHECK(!sequence.isEmpty(), QByteArray(region.length, U2Msa::GAP_CHAR));
+    const char *coreSequence = sequence.constData();
+    int coreLength = sequence.length();
+    QByteArray result;
+    result.reserve(region.length);
+    int gapIndex = 0;  // Current gap index.
+    int corePos = 0;  // Current offset in core sequence.
+    int pos = 0;  // Current gapped position in the row.
+    int regionEnd = (int)region.endPos();  // Last (exclusive) global gapped position to include into the result.
+    if (region.startPos < 0) {
+        result.append(-region.startPos, U2Msa::GAP_CHAR);
+    }
+    // Iterate interleaved gap & core sequence regions.
+    while (gapIndex <= gaps.length() && pos < regionEnd && corePos < coreLength) {
+        const U2MsaGap *gap = gapIndex == gaps.length() ? nullptr : &gaps[gapIndex];
+        bool isValidGapModel = gapIndex == 0 || gap == nullptr || gap->startPos > gaps[gapIndex - 1].endPos();
+        SAFE_POINT(isValidGapModel, "Invalid gap model", {});
+        if (gap == nullptr || gap->startPos > pos) {
+            // Processing core sequence part. At this point no gaps left or the next gap starts after 'pos'.
+            int corePartLength = gap == nullptr ? coreLength - corePos : gap->startPos - pos;
+            SAFE_POINT(corePos + corePartLength <= coreLength, "Invalid position in core sequence!", {});
+            if (region.startPos < pos + corePartLength) {  // Add core sub-sequence to the result.
+                int nChars;
+                int corePartOffset;
+                if (region.startPos <= pos) {  // Add a prefix or whole core sequence region.
+                    nChars = qMin(corePartLength, regionEnd - pos);
+                    corePartOffset = 0;
+                } else if (regionEnd > pos + corePartLength) {  // Add suffix of the core sequence region.
+                    nChars = pos + corePartLength - region.startPos;
+                    corePartOffset = corePartLength - nChars;
+                } else {  // Add middle part of the core sequence region.
+                    SAFE_POINT(region.startPos > pos && regionEnd <= pos + corePartLength, "Invalid middle region coordinates", {};)
```

Comment:
The same

Reply:
Added more memory opts to the fix. Answered all question. Re-checking the tests now